proof/refine/ARM/SchedContextInv_R.thy

```diff
@@ -1268,6 +1268,98 @@ lemma headInsufficientLoop_corres:
   apply (fastforce intro!: non_overlapping_merge_refills_terminates)
   done
 
+lemma refillEmpty_sp:
+  "\<lbrace>P\<rbrace>refillEmpty scp \<lbrace>\<lambda>rv s. P s \<and> (\<forall>ko. ko_at' ko scp s \<longrightarrow> rv = (scRefillCount ko = 0))\<rbrace>"
+  apply (wpsimp wp: refillEmpty_wp)
+  apply (clarsimp simp: obj_at'_def)
+  done
+
+lemma refillFull_sp:
+  "\<lbrace>P\<rbrace> refillFull scp \<lbrace>\<lambda>rv s. P s \<and> (\<forall>ko. ko_at' ko scp s \<longrightarrow> rv = (scRefillCount ko = scRefillMax ko))\<rbrace>"
+  apply (wpsimp wp: refillFull_wp)
+  apply (clarsimp simp: obj_at'_def)
+  done
+
+lemma length_refills_map_is_scRefillCount:
+  "sc_valid_refills' sc
+   \<Longrightarrow> (length (refills_map (scRefillHead sc) (scRefillCount sc) (scRefillMax sc) (scRefills sc)))
+        = scRefillCount sc"
+  apply (clarsimp simp: valid_sched_context'_def refillTailIndex_def refills_map_def)
+  done
```

Comment:
We already have
```
lemma length_refills_map[simp]:
  "⟦ mx ≤ length list; count ≤ mx ⟧ ⟹ length (refills_map start count mx list) = count"
  by (clarsimp simp: refills_map_def)
```
in the simp set. Given that `sc_valid_refills’` is an abbreviation, I don’t see why we need to have a separate lemma for this particular instance.

Reply:
Removed